### ISSUE TITLE
Various false positives when calling dict(existingDict, **kwargs) to make a heterogenous dict

### ISSUE THREAD

**Issue by Wizzerinus** [astral-sh/ty#2586]
### Summary

When a dict is defined with `dict(existingDict, **kwargs)`, ty cannot infer the overload of `dict.__init__` in many cases; seemingly the condition is that the resulting dict would be heterogenous.

For example, with `x = {"x": 1}`, call to `dict(x, z=2)` is accepted, but call to `dict(x, z="str")` cannot be inferred.

When `x` is already heterogenous, nothing can be inferred at all.

```py
A = {"x": 1, "y": (2, 3, 4)}
B = dict(A, z=2)     # this is an error
C = dict(A, z="x")   # this is an error

A1 = {"x": 1}
B1 = dict(A1, z=2)
C1 = dict(A1, z="x")  # this is an error
```

https://play.ty.dev/4beb193f-fe05-4fa2-9b77-ed7fda32aa80

### Version

0.0.13

**Comment by sharifhsn** [comment]
Closed by https://github.com/astral-sh/ruff/pull/21902
